src/coreclr/src/jit/morph.cpp

```diff
@@ -7298,306 +7283,724 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
-        // specially below.
-        fgArgInfo* argInfo = call->fgArgInfo;
-        assert(argInfo != nullptr);
+        call->gtFlags &= ~GTF_CALL_VIRT_STUB;
+    }
+
+    GenTree* callDispatcherAndGetResult = fgCreateCallDispatcherAndGetResult(call, help.hCallTarget, help.hDispatcher);
+
+    // Change the call to a call to the StoreArgs stub.
+    if (call->HasRetBufArg())
+    {
+        JITDUMP("Removing retbuf");
+        call->gtCallArgs = call->gtCallArgs->GetNext();
+        call->gtCallMoreFlags &= ~GTF_CALL_M_RETBUFFARG;
+
+        // We changed args so recompute info.
+        call->fgArgInfo = nullptr;
+    }
 
-        GenTreeCall::Use* vsdArg = nullptr;
+    // If asked to store target and we have a type arg we will store
+    // instantiating stub, so in that case we should not pass the type arg.
+    if ((help.flags & CORINFO_TAILCALL_STORE_TARGET) && call->tailCallInfo->GetSig()->hasTypeArg())
+    {
+        JITDUMP("Removing type arg");
 
-        for (unsigned index = 0; index < argInfo->ArgCount(); ++index)
+        assert(call->gtCallArgs != nullptr);
+        if (Target::g_tgtArgOrder == Target::ARG_ORDER_R2L)
+        {
+            // Generic context is first arg
+            call->gtCallArgs = call->gtCallArgs->GetNext();
+        }
+        else
         {
-            fgArgTabEntry* arg = argInfo->GetArgEntry(index, false);
-            if (arg->isNonStandard)
+            // Generic context is last arg
+            GenTreeCall::Use** lastArgSlot = &call->gtCallArgs;
+            while ((*lastArgSlot)->GetNext() != nullptr)
+                lastArgSlot = &(*lastArgSlot)->NextRef();
+
+            *lastArgSlot = nullptr;
+        }
+        call->fgArgInfo = nullptr;
+    }
+
+    // We may need to pass the target, for instance for calli or generic methods
+    // where we pass instantiating stub.
+    if (help.flags & CORINFO_TAILCALL_STORE_TARGET)
+    {
+        JITDUMP("Adding target since VM requested it\n");
+        GenTree* target;
+        if (call->tailCallInfo->IsCalli())
+        {
+            noway_assert(call->gtCallType == CT_INDIRECT && call->gtCallAddr != nullptr);
+            target = call->gtCallAddr;
+        }
+        else
+        {
+            CORINFO_CALL_INFO callInfo;
+            unsigned          flags = CORINFO_CALLINFO_LDFTN;
+            if (call->tailCallInfo->IsCallvirt())
             {
-                // The only supported nonstandard arg for slow tailcalls is
-                // VSD arg.
-                assert(vsdArg == nullptr);
-                vsdArg = arg->use;
-#ifndef DEBUG
-                break;
-#endif
+                flags |= CORINFO_CALLINFO_CALLVIRT;
+            }
+
+            eeGetCallInfo(call->tailCallInfo->GetToken(), nullptr, (CORINFO_CALLINFO_FLAGS)flags, &callInfo);
+
+            // R2R requires different handling but we don't support tailcall via
+            // helpers in R2R yet, so just leave it for now.
+            assert(!opts.IsReadyToRun());
```

Comment:
Please add `TODO: R2R: TailCallViaHelper`

Reply:
Seems like this assert/comment should be higher up? Unless I'm confused, we should never even call `fgMorphTailCallViaHelpers` when compiling for R2R.